File: lib/go/registry.go
package frugal

import (
	"fmt"
	"strconv"
	"sync"

	"git.apache.org/thrift.git/lib/go/thrift"
)

// FAsyncCallback is an internal callback which is constructed by generated
// code and invoked by an FRegistry when a RPC response is received. In other
// words, it's used to complete RPCs. The operation ID on FContext is used to
// look up the appropriate callback. FAsyncCallback is passed an in-memory
// TTransport which wraps the complete message. The callback returns an error
// or throws an exception if an unrecoverable error occurs and the transport
// needs to be shutdown.
type FAsyncCallback func(thrift.TTransport) error

// FRegistry is responsible for multiplexing and handling received messages.
// Typically there is a client implementation and a server implementation. An
// FRegistry is used by an FTransport.
//
// The client implementation is used on the client side, which is making RPCs.
// When a request is made, an FAsyncCallback is registered to an FContext. When
// a response for the FContext is received, the FAsyncCallback is looked up,
// executed, and unregistered.
//
// The server implementation is used on the server side, which is handling
// RPCs. It does not actually register FAsyncCallbacks but rather has an
// FProcessor registered with it. When a message is received, it's buffered and
// passed to the FProcessor to be handled.
type fRegistry interface {
	// Register a channel for the given Context.
	Register(ctx FContext, resultC chan []byte) error
	// Unregister a callback for the given Context.
	Unregister(FContext)
	// Execute dispatches a single Thrift message frame.
	Execute([]byte) error
}

type fRegistryImpl struct {
	mu       sync.RWMutex
	channels map[uint64]chan []byte
}

// NewFRegistry creates a Registry intended for use by Frugal clients.
// This is only to be called by generated code.
func newFRegistry() fRegistry {
	return &fRegistryImpl{channels: make(map[uint64]chan []byte)}
}

// Register a channel for the given Context.
func (c *fRegistryImpl) Register(ctx FContext, resultC chan []byte) error {
	// An FContext can be reused for multiple requests. Because of this,
	// FContext's have a monotonically increasing atomic uint64. We check
	// the channels map to ensure that request is not still in-flight.
	opID, err := getOpID(ctx)

	c.mu.Lock()
	defer c.mu.Unlock()
	if err == nil {
		_, ok := c.channels[opID]
		if ok {
			return fmt.Errorf("frugal: context already registered, opid %d is in-flight for another request", opID)
		}
	}
	c.channels[opID] = resultC
	return nil
}

// Unregister a callback for the given Context.
func (c *fRegistryImpl) Unregister(ctx FContext) {
	opID, err := getOpID(ctx)
	if err != nil {
		logger().Warnf("Attempted to unregister an FContext with a malformed opid: %s", err)
		return
	}
	c.mu.Lock()
	delete(c.channels, opID)
	c.mu.Unlock()
}

// Execute dispatches a single Thrift message frame.
func (c *fRegistryImpl) Execute(frame []byte) error {
	headers, err := getHeadersFromFrame(frame)
	if err != nil {
		logger().Warn("frugal: invalid protocol frame headers:", err)
		return err
	}

	opid, err := strconv.ParseUint(headers[opIDHeader], 10, 64)
	if err != nil {
		logger().Warn("frugal: invalid protocol frame, op id not a uint64:", err)
		return err
	}

	c.mu.RLock()
	resultC, ok := c.channels[opid]
	if !ok {
		logger().Warn("frugal: unregistered context")
		c.mu.RUnlock()
		return nil
	}
	c.mu.RUnlock()

	resultC <- frame
	return nil
}

File: lib/python/frugal/registry.py
import logging
from threading import Lock

from thrift.transport.TTransport import TMemoryBuffer

from frugal.context import _OPID_HEADER
from frugal.exceptions import FException
from frugal.util.headers import _Headers

logger = logging.getLogger(__name__)


class FRegistry(object):
    """
    Registry is responsible for multiplexing received
    messages to the appropriate callback.
    """

    def register(self, context, callback):
        """Register a callback for a given FContext.

        Args:
            context: FContext to register.
            callback: function to register.
        """
        pass

    def unregister(self, context):
        """Unregister the callback for a given FContext.

        Args:
            context: FContext to unregister.
        """
        pass

    def execute(self, frame):
        """Dispatch a single Frugal message frame.

        Args:
            frame: an entire Frugal message frame.
        """
        pass


class FRegistryImpl(FRegistry):
    """
    FRegistryImpl is intended for use only by Frugal clients.
    This is only to be used by generated code.
    """

    def __init__(self):
        self._handlers = {}
        self._handlers_lock = Lock()
        self._next_opid = 0
        self._opid_lock = Lock()

    def register(self, context, callback):
        """Register a callback for a given FContext.

        Args:
            context: FContext to register.
            callback: function to register.
        """
        # An FContext can be reused for multiple requests. Because of this,
        # every time an FContext is registered, it must be assigned a new op id
        # to ensure we can properly correlate responses. We use a monotonically
        # increasing atomic uint64 for this purpose. If the FContext already
        # has an op id, it has been used for a request. We check the handlers
        # map to ensure that request is not still in-flight.
        with self._handlers_lock:
            if str(context._get_op_id()) in self._handlers:
                raise FException("context already registered")

        op_id = self._increment_and_get_next_op_id()
        context._set_op_id(op_id)

        with self._handlers_lock:
            self._handlers[str(op_id)] = callback

    def unregister(self, context):
        """Unregister the callback for a given FContext.

        Args:
            context: FContext to unregister.
        """
        with self._handlers_lock:
            self._handlers.pop(str(context._get_op_id()), None)

    def execute(self, frame):
        """Dispatch a single Frugal message frame.

        Args:
            frame: an entire Frugal message frame.
        """
        headers = _Headers.decode_from_frame(frame)
        op_id = headers.get(_OPID_HEADER, None)

        if not op_id:
            raise FException("Frame missing op_id")

        with self._handlers_lock:
            handler = self._handlers.get(op_id, None)
            if not handler:
                return

            handler(TMemoryBuffer(frame))

    def _increment_and_get_next_op_id(self):
        with self._opid_lock:
            self._next_opid += 1
            op_id = self._next_opid
        return op_id
